crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,397 @@
+/// - `[<prefix>/]<partition_id>/{lsn}_{snapshot_id}/*.sst` - data files (explicitly named in `metadata.json`)
+#[derive(Clone)]
+pub struct SnapshotRepository {
+    object_store: Arc<dyn ObjectStore>,
+    destination: Url,
+    prefix: String,
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LatestSnapshot {
+    pub version: SnapshotFormatVersion,
+
+    pub partition_id: PartitionId,
+
+    /// Restate cluster name which produced the snapshot.
+    pub cluster_name: String,
+
+    /// Node that produced this snapshot.
+    pub node_name: String,
+
+    /// Local node time when the snapshot was created.
+    #[serde(with = "serde_with::As::<serde_with::DisplayFromStr>")]
+    pub created_at: humantime::Timestamp,
+
+    /// Unique snapshot id.
+    pub snapshot_id: SnapshotId,
+
+    /// The minimum LSN guaranteed to be applied in this snapshot. The actual
+    /// LSN may be >= [minimum_lsn].
+    pub min_applied_lsn: Lsn,
+
+    /// The relative path within the snapshot repository where the snapshot data is stored.
+    pub path: String,
+}
+
+impl SnapshotRepository {
+    pub async fn create(
+        base_dir: PathBuf,
+        snapshots_options: &SnapshotsOptions,
+    ) -> anyhow::Result<SnapshotRepository> {
+        let destination = if let Some(ref destination) = snapshots_options.destination {
+            destination.clone()
+        } else {
+            base_dir
+                .join("pp-snapshots")
+                .into_os_string()
+                .into_string()
+                .map(|path| format!("file://{path}"))
+                .map_err(|e| anyhow!("Unable to convert path to string: {:?}", e))?
+        };
+        let destination =
+            Url::parse(&destination).context("Failed parsing snapshot repository URL")?;
+
+        // AWS-specific ergonomics optimization: without explicit configuration, we set up the AWS
+        // SDK credentials provider so that the conventional environment variables and config
+        // locations just work. This makes object_store behave similarly to the Lambda invoker.
+        let object_store: Arc<dyn ObjectStore> = if destination.scheme() == "s3"
+            && destination.query().is_none()
+            && snapshots_options.additional_options.is_empty()
+        {
+            debug!("Using AWS SDK credentials provider");
+            let aws_region = aws_config::load_defaults(BehaviorVersion::v2024_03_28())
+                .await
+                .region()
+                .context("Unable to determine AWS region to use with S3")?
+                .clone();
+
+            let store = AmazonS3Builder::new()
+                .with_url(destination.clone())
+                .with_region(aws_region.to_string())
+                .with_credentials(Arc::new(AwsSdkCredentialsProvider {
+                    credentials_provider: DefaultCredentialsChain::builder().build().await,
+                }))
```

Comment:
What's the difference to `AmazonS3Builder::from_env()`?

Reply:
The comment above tries to explain, but doesn't go into detail. `AmazonS3Builder::from_env()` only respects a handful of OS environment variables, and supports limited session credentials. Using the AWS SDK's configuration mechanism has a whole chain of potential config sources, from environment variables to various config files locations that AWS CLIs and SDKs commonly use. It also has support for refreshing session creds obtained from an AWS profile, such as what we use with our SSO login - and are very common for development machines. This is not only more ergonomic, but also more secure as session creds only ever exist in process memory, and not in OS variables.

Reply:
I expanded the comment - let me know if that would have helped?

Reply:
Thanks for the clarification Pavel!